SIPS/sip-20.md

```diff
@@ -2,188 +2,213 @@
- * entry point, as described in SIP-21.
- * @property connections - The external data connections.
- */
-type ExternalData = {
-    maxRequestTime?: number;
-    connections: ExternalDataConnection[]
+```typescript
+type CloseWebSocketParams = {
+  id: string;
 };
 ```
+The RPC method takes one parameter:
+- `id` - The unique identifier of the WebSocket connection to close. This identifier is returned by the `snap_openWebSocket` method.
 
-#### Example
-
-The new field can be specified as follows in a `snap.manifest.json` file:
-
-```json
-{
-  "initialPermissions": {
-    "endowment:external-data": {
-      "maxRequestTime": 50000,
-      "connections": [
-        {
-          "type": "websocket",
-          "url": "wss://example.com/binary/endpoint",
-          "dataType": "binary"
-        },
-        {
-          "type": "websocket",
-          "url": "wss://example.com/text/endpoint",
-          "dataType": "text"
-        }
-      ]
-    }
-  }
-}
+An example of usage is given below.
+
+```typescript
+snap.request({
+  method: "snap_closeWebSocket",
+  params: {
+    id: "unique-connection-id",
+  },
+});
 ```
+#### `snap_sendMessage`
+This method allows a Snap to send a message over an existing WebSocket connection. The method takes the unique identifier of the connection and the message to send as parameters.
 
-### Permission caveats
+```typescript
+type SendWebSocketMessageParams = {
+  id: string;
+  message: string | Uint8Array;
+};
+```
 
-Each caveat object MUST have `type` property that SHOULD be `websocket`.
-The caveat MUST have `url` and `dataType` properties.
-The `url` MUST be string.
-The `dataType` MUST be string that MUST be `"binary"` or `"text"`.
+The RPC method takes two parameters:
+- `id` - The unique identifier of the WebSocket connection to send the message over. This identifier is returned by the `snap_openWebSocket` method.
+- `message` - The message to send over the WebSocket connection. It can be either a string or a `Uint8Array`.
 
-- `type` property represents the type of data source.
-- `url` is the WebSocket URL.
-- `dataType` is type of data that WebSocket returns.
+An example of usage is given below.
 
-The caveats MUST NOT have duplicate objects with the same `url` properties.
+```typescript
+snap.request({
+  method: "snap_sendMessage",
+  params: {
+    id: "unique-connection-id",
+    message: "Hello, WebSocket!",
+  },
+});
+```
+#### `snap_getWebSockets`
+This method allows a Snap to retrieve a list of all currently open WebSocket connections. It returns an array of objects, each containing the unique identifier and URL of the connection.
 
-The `url` MUST start with `wss://` which is a protocol indicator for secure WebSocket connection.
+- `id` - The unique identifier of the WebSocket connection.
+- `url` - The URL of the WebSocket connection.
 
-### Snap implementation
+```typescript
+type WebSocketConnection = {
+  id: string;
+  url: string;
```

Comment:
We should choose either URL or origin and use that everywhere

Reply:
I agree that for open connections `origin` is the way to go but to open a connection like in `snap_openWebSocket` we should use `url` as `origin` doesn't make sense